File: SentinelAI-v2-main/sentinelai-v2/backend/agents/runtime_agent.py
"""
SentinelAI v2.0 — Runtime Agent
Model: Phi-4-mini-reasoning (shared instance with Content Agent)
Analyzes real-time hook events: fetch, WebSocket, DOM mutations, cookie access, eval
"""

MODEL_NAME = "phi4-mini"


def analyze_heuristic(events: list, hostname: str = "") -> dict:
    """Heuristic runtime event analysis."""
    threats = []
    score = 0

    if not events:
        return {"agent": "runtime-agent", "model": MODEL_NAME, "score": 0, "threats": []}

    hook_counts = {}
    for e in events:
        hook = e.get("hook", "unknown")
        hook_counts[hook] = hook_counts.get(hook, 0) + 1

    # Excessive network calls
    net_calls = sum(hook_counts.get(h, 0) for h in ['fetch', 'xhr', 'beacon'])
    if net_calls > 20:
        score += 15
        threats.append({"type": "excessive-network", "detail": f"{net_calls} outbound network calls"})

    # Eval usage
    if hook_counts.get('eval', 0) > 0:
        score += 20
        threats.append({"type": "eval-usage", "detail": f"{hook_counts['eval']} eval/Function calls"})
        for e in events:
            if e.get("hook") == "eval" and e.get("data", {}).get("hasBase64"):
                score += 15
                threats.append({"type": "eval-base64", "detail": "Base64 decoding in eval (obfuscated malware)"})

    # DOM injection
    for e in events:
        if e.get("hook") == "mutation":
            reason = e.get("data", {}).get("reason", "")
            if reason == "overlay-injection":
                score += 25
                threats.append({"type": "overlay-injection", "detail": "Full-screen overlay (clickjacking)"})
            elif reason == "iframe-injection":
                score += 20
                threats.append({"type": "iframe-injection", "detail": "Hidden iframe injected"})
            elif reason == "form-injection":
                score += 20
                threats.append({"type": "form-injection", "detail": "Form dynamically injected"})

    # Cookie theft
    cookie_reads = [e for e in events if e.get("hook") == "cookie" and e.get("data", {}).get("action") == "read"]
    session_reads = [e for e in cookie_reads if e.get("data", {}).get("hasSessionId")]
    if session_reads:
        score += 20
        threats.append({"type": "session-theft", "detail": f"Session cookies accessed {len(session_reads)} times"})

    # Permission abuse
    perm_events = [e for e in events if e.get("hook") == "permission"]
    for pe in perm_events:
        action = pe.get("data", {}).get("action", "")
        if action == "getUserMedia":
            score += 20
            media = []
            if pe.get("data", {}).get("video"):
                media.append("camera")
            if pe.get("data", {}).get("audio"):
                media.append("microphone")
            threats.append({"type": "media-access", "detail": f"Page requested {'+'.join(media)} access"})
        elif action == "geolocation":
            score += 10
            threats.append({"type": "geolocation-access", "detail": "Page requested geolocation"})

    # WebSocket C2 channels
    ws_connects = [e for e in events if e.get("hook") == "websocket" and e.get("data", {}).get("action") == "connect"]
    if len(ws_connects) > 3:
        score += 15
        threats.append({"type": "c2-channel", "detail": f"{len(ws_connects)} WebSocket connections (possible C2)"})

    # Canvas fingerprinting
    if hook_counts.get('canvas', 0) > 0:
        score += 10
        threats.append({"type": "fingerprinting", "detail": "Canvas/WebGL fingerprinting detected"})

    # Clipboard access
    clip_reads = [e for e in events if e.get("hook") == "clipboard" and e.get("data", {}).get("action") == "read"]
    if clip_reads:
        score += 15
        threats.append({"type": "clipboard-read", "detail": "Page read clipboard contents"})

    return {
        "agent": "runtime-agent",
        "model": MODEL_NAME,
        "score": min(score, 100),
        "threats": threats,
        "hook_counts": hook_counts
    }


async def analyze_llm(events: list, hostname: str = "") -> dict:
    """Enhanced analysis using Phi-3.5-mini-instruct via AirLLM (shared instance)."""
    heuristic = analyze_heuristic(events, hostname)

    event_summary = f"Hook counts: {heuristic.get('hook_counts', {})}\n"
    event_summary += f"Hostname: {hostname}\n"
    event_summary += f"Sample events: {events[:5]}"

    prompt = f"""You are a runtime behavior analysis agent. Analyze these browser runtime events for threats.

{event_summary}

Heuristic findings: {heuristic['threats']}

Provide JSON with:
- "behavior_assessment": brief analysis of what the page is doing
- "is_malicious": true/false
- "confidence": 0-100

Respond ONLY with valid JSON."""

    try:
        from backend.ollama_engine import generate_async, is_available
        if is_available():
            heuristic["llm_analysis"] = await generate_async("runtime-agent", prompt)
        else:
            heuristic["llm_analysis"] = "AirLLM not installed"
    except Exception as e:
        heuristic["llm_analysis"] = f"LLM unavailable: {str(e)}"

    return heuristic

File: SentinelAI-v2-main/sentinelai-v2/agents/runtime-agent.js
/**
 * SentinelAI v2.0 — Runtime Agent
 * Analyzes runtime hook events for threat signals.
 * Stub Phi-4-mini LLM call for advanced behavioral analysis.
 */

class RuntimeAgent {
  constructor() {
    this.name = 'runtime-agent';
  }

  /**
   * Analyze a batch of hook events
   * @param {Array} events - Hook events from content script
   * @param {string} hostname - The page hostname
   * @returns {Object} - { score: 0-100, threats: [] }
   */
  analyze(events, hostname) {
    const threats = [];
    let score = 0;

    if (!events || events.length === 0) {
      return { agent: this.name, score: 0, threats: [], timestamp: Date.now() };
    }

    // Categorize events by hook
    const hookCounts = {};
    events.forEach(e => {
      hookCounts[e.hook] = (hookCounts[e.hook] || 0) + 1;
    });

    // 1. Excessive network calls (potential exfiltration)
    const networkCalls = (hookCounts['fetch'] || 0) + (hookCounts['xhr'] || 0) + (hookCounts['beacon'] || 0);
    if (networkCalls > 20) {
      score += 15;
      threats.push({ type: 'excessive-network', detail: `${networkCalls} outbound network calls detected` });
    }

    // 2. Eval usage (code injection)
    if (hookCounts['eval'] > 0) {
      score += 20;
      threats.push({ type: 'eval-usage', detail: `${hookCounts['eval']} eval/Function calls detected` });

      // Check eval events for suspicious patterns
      events.filter(e => e.hook === 'eval').forEach(e => {
        if (e.data?.hasBase64) {
          score += 15;
          threats.push({ type: 'eval-base64', detail: 'Eval contains base64 decoding (possible obfuscated malware)' });
        }
        if (e.data?.hasNetworkCall) {
          score += 15;
          threats.push({ type: 'eval-network', detail: 'Eval contains network calls (possible exfiltration)' });
        }
      });
    }

    // 3. DOM injection (mutation observer events)
    const mutationEvents = events.filter(e => e.hook === 'mutation');
    const iframeInjections = mutationEvents.filter(e => e.data?.reason === 'iframe-injection');
    const overlayInjections = mutationEvents.filter(e => e.data?.reason === 'overlay-injection');
    const formInjections = mutationEvents.filter(e => e.data?.reason === 'form-injection');

    if (iframeInjections.length > 0) {
      score += 20;
      threats.push({ type: 'iframe-injection', detail: `${iframeInjections.length} iframe(s) dynamically injected` });
    }
    if (overlayInjections.length > 0) {
      score += 25;
      threats.push({ type: 'overlay-injection', detail: 'Full-screen overlay detected (possible clickjacking)' });
    }
    if (formInjections.length > 0) {
      score += 20;
      threats.push({ type: 'form-injection', detail: 'Form dynamically injected (possible credential phishing)' });
    }

    // 4. DOM write with scripts
    const domWriteEvents = events.filter(e => e.hook === 'dom-write');
    domWriteEvents.forEach(e => {
      if (e.data?.hasScript) {
        score += 15;
        threats.push({ type: 'script-injection', detail: 'Script injected via document.write/innerHTML' });
      }
    });

    // 5. WebSocket connections
    const wsConnections = events.filter(e => e.hook === 'websocket' && e.data?.action === 'connect');
    if (wsConnections.length > 3) {
      score += 10;
      threats.push({ type: 'excessive-websocket', detail: `${wsConnections.length} WebSocket connections opened` });
    }

    // 6. Canvas fingerprinting
    if (hookCounts['canvas'] > 0) {
      score += 10;
      threats.push({ type: 'fingerprinting', detail: 'Browser fingerprinting via canvas/WebGL detected' });
    }

    // 7. Clipboard access
    const clipboardReads = events.filter(e => e.hook === 'clipboard' && e.data?.action === 'read');
    if (clipboardReads.length > 0) {
      score += 15;
      threats.push({ type: 'clipboard-read', detail: 'Page attempted to read clipboard' });
    }

    // 8. Deferred suspicious timers
    const suspiciousTimers = events.filter(e => e.hook === 'timer' && e.data?.hasSuspiciousPatterns);
    if (suspiciousTimers.length > 0) {
      score += 15;
      threats.push({ type: 'deferred-malware', detail: `${suspiciousTimers.length} deferred suspicious timer(s) detected` });
    }

    // 9. Cross-origin form submissions
    const formEvents = events.filter(e => e.hook === 'form' && e.data?.actionIsCrossOrigin);
    if (formEvents.length > 0) {
      formEvents.forEach(e => {
        if (e.data?.hasPasswordField) {
          score += 30;
          threats.push({ type: 'credential-exfil', detail: 'Password submitted to cross-origin server' });
        }
      });
    }

    // 10. Beacon exfiltration patterns
    const beaconEvents = events.filter(e => e.hook === 'beacon');
    if (beaconEvents.length > 5) {
      score += 15;
      threats.push({ type: 'beacon-exfil', detail: `${beaconEvents.length} sendBeacon calls (possible tracking/exfiltration)` });
    }

    return {
      agent: this.name,
      score: Math.min(score, 100),
      threats,
      hookCounts,
      timestamp: Date.now()
    };
  }

  /**
   * Stub LLM call — would call Phi-4-mini for behavioral analysis
   */
  async analyzeLLM(events, hostname) {
    return this.analyze(events, hostname);
  }
}

if (typeof globalThis !== 'undefined') {
  globalThis.SentinelRuntimeAgent = RuntimeAgent;
}
